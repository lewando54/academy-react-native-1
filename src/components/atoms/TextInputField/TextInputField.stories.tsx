/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { StoryObj, Meta } from '@storybook/react'
import { useState } from 'react'

import TextInputField from './TextInputField'
import { create } from 'react-test-renderer'

interface IInputMockupProps {
    placeholder: string,
    value: string,
    secureTextEntry: boolean
}

function StatefulTextInputField({placeholder, value, secureTextEntry}: IInputMockupProps){
    const [textValue, setTextValue] = useState(value)

    return (
        <TextInputField value={textValue} placeholder={placeholder} secureTextEntry={secureTextEntry} onChange={(newValue) => setTextValue(newValue)}></TextInputField>
    )
}

const meta = {
    component: StatefulTextInputField,
    argTypes: {
        placeholder: { control: 'text' },
        value: { control: 'text' },
        secureTextEntry: { control: 'boolean' }
    }
} satisfies Meta<typeof StatefulTextInputField>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
    args: {
        placeholder: 'Type something',
        value: 'Initial value',
        secureTextEntry: false,
    }
}