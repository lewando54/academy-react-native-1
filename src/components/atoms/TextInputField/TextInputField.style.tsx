import { StyleSheet } from 'react-native'

export default StyleSheet.create({
    base:{
        width: '100%',
        borderWidth: 2,
        borderColor: '#e5e7eb',
        backgroundColor: '#fff',
        borderRadius: 6,
        fontSize: 18,
        lineHeight: 28,
        padding: 4,
        color: 'black'
    },
    focused: {
        borderColor: '#689ff7'
    }
})