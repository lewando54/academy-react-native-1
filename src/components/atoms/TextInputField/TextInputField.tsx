import React, { useState, useRef, forwardRef, useImperativeHandle } from 'react'
import { TextInput } from 'react-native'
import textInputFieldStyle from './TextInputField.style'

export interface ITextInputFieldRef {
    focus: () => void
}
export interface IInputProps {
    name?: string
    required?: boolean
    testId?: string
    placeholder?: string
    value: string
    secureTextEntry?: boolean
    onChange: (newValue: string) => void
}

function TextInputField({name, required, testId, placeholder, secureTextEntry=false, value='', onChange}: IInputProps, ref: React.Ref<ITextInputFieldRef>): React.ReactElement{
    const [isFocused, setIsFocused] = useState(false)

    const TextInputRef = useRef<TextInput>(null)

    useImperativeHandle(ref, () => {
        return {
            focus() {
                TextInputRef.current?.focus()
            }
        }
    })

    const onFocus = () => {
        setIsFocused((prev) => !prev)
    }

    const onBlurred = () => {
        setIsFocused((prev) => !prev)
    }

    const handleOnChange = (newValue: string) => {
        if(onChange){
            onChange(newValue)
        }
    }

    return (
        <TextInput value={value}
        placeholder={placeholder}
        onChangeText={handleOnChange}
        onFocus={onFocus}
        onBlur={onBlurred}
        testID={testId}
        ref={TextInputRef}
        secureTextEntry={secureTextEntry}
        style={[
            textInputFieldStyle.base,
            isFocused && textInputFieldStyle.focused
        ]}/>
    )
}

export default forwardRef(TextInputField)