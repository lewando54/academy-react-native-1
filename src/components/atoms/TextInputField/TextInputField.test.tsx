import React, { createRef, useState, forwardRef } from 'react'
import { render, screen, fireEvent, waitFor } from '@testing-library/react-native'
import * as renderer from 'react-test-renderer'
import TextInputField, { ITextInputFieldRef } from './TextInputField';
import textInputFieldStyle from './TextInputField.style'

interface ITextInputFieldMockProps{
    initialText?: string
    placeholder?: string
    secureTextEntry?: boolean
}

const StatefulTextInputField = forwardRef(({initialText='', placeholder='', secureTextEntry=false}: ITextInputFieldMockProps, ref: React.Ref<ITextInputFieldRef>) => {
    const [text, setText] = useState(initialText)

    return(
        <TextInputField value={text} placeholder={placeholder} secureTextEntry={false} testId='textinput-test' ref={ref} onChange={(newValue) => setText(newValue)}/>
    )
})

describe('TextInputField', () => {
    it('should render correctly', () => {
        const tree = renderer.create(<TextInputField value='test' onChange={() => {}} />)
        expect(tree).toMatchSnapshot()
    })

    it('shoud focus when focus from ref is called', () => {
        const textInputFieldRef = createRef<ITextInputFieldRef>()
        const { getByTestId } = render(<StatefulTextInputField ref={textInputFieldRef}/>)
        textInputFieldRef.current.focus()
        waitFor(() => {
            expect(getByTestId('textinput-test').props.style).toContain(textInputFieldStyle.focused)
        })
    })

    it('should input text correctly', () => {
        const { getByTestId, getByDisplayValue } = render(<StatefulTextInputField/>)
        fireEvent.changeText(getByTestId('textinput-test'), 'Hello world!')
        expect(getByDisplayValue('Hello world!')).toBeTruthy()
    })

    it('should apply focused style when input is focused and remove it when blurred', () => {
        const { getByTestId } = render(<StatefulTextInputField/>)

        const input = getByTestId('textinput-test')
        expect(input.props.style).not.toContain(textInputFieldStyle.focused)

        fireEvent(input, 'focus')
        expect(input.props.style).toContain(textInputFieldStyle.focused)

        fireEvent(input, 'blur')
        expect(input.props.style).not.toContain(textInputFieldStyle.focused)
    })

    it('should not call onChange when it is not passed as a prop', () => {
        const { getByTestId } = render(
            <TextInputField 
                value={'text'} 
                placeholder={'placeholder'} 
                secureTextEntry={false} 
                testId='textinput-test'
                onChange={undefined}
            />
        )
        const input = getByTestId('textinput-test')
        fireEvent.changeText(input, 'new value')
    })
})