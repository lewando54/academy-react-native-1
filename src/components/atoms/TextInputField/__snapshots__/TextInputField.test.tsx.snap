// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputField should render correctly 1`] = `
<TextInput
  onBlur={[Function]}
  onChangeText={[Function]}
  onFocus={[Function]}
  secureTextEntry={false}
  style={
    [
      {
        "backgroundColor": "#fff",
        "borderColor": "#e5e7eb",
        "borderRadius": 6,
        "borderWidth": 2,
        "color": "black",
        "fontSize": 18,
        "lineHeight": 28,
        "padding": 4,
        "width": "100%",
      },
      false,
    ]
  }
  value="test"
/>
`;
